File: backend/app/schemas/ops.py
"""포털 운영 설정 스키마 (api.md §9, A-OP-01·02·03·04)."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


# --- A-OP-04 시스템 설정 ---------------------------------------------------
class SettingsOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    #: C-04 — 실시간성 요구가 30초 이하다.
    poll_interval_sec: int
    session_timeout_min: int
    smtp_host: str | None
    smtp_port: int | None
    smtp_sender: str | None
    webhook_url: str | None
    updated_at: datetime | None = None


class SettingsUpdate(BaseModel):
    poll_interval_sec: int | None = Field(default=None, ge=5, le=30)
    session_timeout_min: int | None = Field(default=None, ge=5, le=10080)
    smtp_host: str | None = None
    smtp_port: int | None = Field(default=None, ge=1, le=65535)
    smtp_sender: str | None = None
    webhook_url: str | None = None


# --- A-OP-01 공지 -----------------------------------------------------------
class NoticeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    body: str | None
    banner_enabled: bool
    start_at: datetime | None
    end_at: datetime | None
    created_at: datetime


class NoticeCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    body: str | None = None
    banner_enabled: bool = False
    start_at: datetime | None = None
    end_at: datetime | None = None


class NoticeUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    body: str | None = None
    banner_enabled: bool | None = None
    start_at: datetime | None = None
    end_at: datetime | None = None


# --- A-OP-03 감사 로그 ------------------------------------------------------
class AuditLogOut(BaseModel):
    id: int
    at: datetime
    #: 표시용 이름. GUID는 화면에서 쓸모가 없어 서버가 바꿔 준다.
    actor: str | None
    actor_role: str | None
    action: str
    cluster_id: int | None
    target: str | None
    detail: str | None
    ip: str | None


# --- A-OP-02 앱 카탈로그 ----------------------------------------------------
APP_KIND = "^(interactive|batch)$"

class AppCatalogOut(BaseModel):
    """코드 카탈로그의 앱과 등록된 메타데이터를 겹친 한 줄."""

    model_config = ConfigDict(from_attributes=True)

    #: 등록 행의 PK. **None이면 아직 등록되지 않은 코드 앱**이다(수정이 아니라 등록 대상).
    id: int | None
    #: 코드 카탈로그(`session_apps`·`batch_apps`)에 있는 앱인가.
    in_code: bool = False
    kind: str
    app_id: str
    name: str
    vendor: str | None
    version: str | None
    image_file: str | None
    #: 이미지의 출처(OCI 참조). **어디서 왔는지**를 아는 유일한 자리다.
    image_ref: str | None = None
    #: 아이콘 파일명(등록 폼이 고르는 값). 파일이 없으면 None.
    icon_file: str | None = None
    #: 화면이 그대로 <img src>에 넣는 주소. 서버가 파일명으로 만든다.
    icon_url: str | None = None
    description: str | None
    updated_at: datetime | None = None


class AppCatalogCreate(BaseModel):
    kind: str = Field(..., pattern=APP_KIND)
    app_id: str = Field(..., min_length=1, max_length=32)
    name: str = Field(..., min_length=1, max_length=128)
    vendor: str | None = Field(default=None, max_length=128)
    version: str | None = Field(default=None, max_length=32)
    image_file: str | None = Field(default=None, max_length=128)
    #: 모양 검사는 **서비스**가 한다(`app_images.IMAGE_REF`) — 정규식 정본이 거기 있고,
    #: 스키마가 서비스를 import하는 전례를 만들지 않는다.
    image_ref: str | None = Field(default=None, max_length=255)
    icon_file: str | None = Field(default=None, max_length=64)
    description: str | None = None


class AppCatalogUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    vendor: str | None = Field(default=None, max_length=128)
    version: str | None = Field(default=None, max_length=32)
    image_file: str | None = Field(default=None, max_length=128)
    #: 모양 검사는 **서비스**가 한다(`app_images.IMAGE_REF`) — 정규식 정본이 거기 있고,
    #: 스키마가 서비스를 import하는 전례를 만들지 않는다.
    image_ref: str | None = Field(default=None, max_length=255)
    icon_file: str | None = Field(default=None, max_length=64)
    description: str | None = None

File: frontend/src/api/ops.ts
import { api, csrfHeaders, PortalApiError } from './client'
import type { OkResponse, Page } from '@/types/api'

/** 포털 설정 (A-OP-04). 폴링 주기는 C-04로 30초 이하가 요구된다. */
export interface Settings {
  poll_interval_sec: number
  session_timeout_min: number
  smtp_host: string | null
  smtp_port: number | null
  smtp_sender: string | null
  webhook_url: string | null
  updated_at: string | null
}

export interface Notice {
  id: number
  title: string
  body: string | null
  banner_enabled: boolean
  start_at: string | null
  end_at: string | null
  created_at: string
}
/**
 * 앱 카탈로그 (A-OP-02). 앱의 **정보**만 담는다 — 실행 방식(기동 스크립트·transport·
 * 파라미터 스키마)은 백엔드 코드 카탈로그가 정본이고 여기로 오지 않는다.
 * `kind` + `app_id`가 코드 카탈로그로 가는 연결 키다.
 */
export interface AppCatalog {
  /** 등록 행의 PK. **null이면 아직 등록되지 않은 코드 앱**(수정이 아니라 등록 대상). */
  id: number | null
  /** 코드 카탈로그에 있는 앱인가. false = 관리자가 미리 등록해 둔 도입 예정 앱. */
  in_code: boolean
  kind: 'interactive' | 'batch'
  app_id: string
  name: string
  vendor: string | null
  version: string | null
  /** 컨테이너 이미지 **파일명**. 파일은 클러스터의 공유 이미지 디렉터리에 있다. */
  image_file: string | null
  /** 아이콘 **파일명**. 등록 폼이 고르는 값이며 서버의 아이콘 디렉터리에 있는 파일이다. */
  icon_file: string | null
  /** 서버가 파일명으로 만든 주소. 화면은 이걸 그대로 <img src>에 넣는다. */
  icon_url: string | null
  description: string | null
  updated_at: string | null
}

/** 감사 로그 (A-OP-03, C-05). actor는 서버가 GUID를 이름으로 바꿔 준다. */
export interface AuditLog {
  id: number
  at: string
  actor: string | null
  actor_role: string | null
  action: string
  cluster_id: number | null
  target: string | null
  detail: string | null
  ip: string | null
}

export const opsApi = {
  settings: () => api.get<Settings>('/settings'),
  updateSettings: (payload: Partial<Settings>) => api.put<Settings>('/settings', payload),

  /** 공지는 **포털 전체 대상**이다 — 클러스터로 거르지 않는다. */
  notices: (params?: { banner?: boolean }) =>
    api.get<Notice[]>(`/notices${params?.banner ? '?banner=true' : ''}`),
  createNotice: (payload: Partial<Notice>) => api.post<Notice>('/notices', payload),
  updateNotice: (id: number, payload: Partial<Notice>) =>
    api.patch<Notice>(`/notices/${id}`, payload),
  removeNotice: (id: number) => api.del<OkResponse>(`/notices/${id}`),

  /** 앱 카탈로그 (A-OP-02). 읽기는 인증 사용자, 쓰기는 ADMIN. */
  apps: () => api.get<AppCatalog[]>('/apps'),
  createApp: (payload: Partial<AppCatalog>) => api.post<AppCatalog>('/apps', payload),
  updateApp: (id: number, payload: Partial<AppCatalog>) =>
    api.patch<AppCatalog>(`/apps/${id}`, payload),
  removeApp: (id: number) => api.del<OkResponse>(`/apps/${id}`),
  /** 아이콘 디렉터리에 놓인 파일명 목록 — 등록 폼의 선택지. */
  appIcons: () => api.get<string[]>('/app-icons'),
  /** 컨테이너 이미지 디렉터리의 파일명 목록. */

  /** 아이콘 업로드. 응답은 갱신된 파일명 목록이며, **이름은 서버가 정한다**. */
  async uploadAppIcon(file: File): Promise<string[]> {
    const form = new FormData()
    form.append('file', file)
    // Content-Type은 지정하지 않는다 — 브라우저가 multipart 경계를 붙여야 한다.
    const response = await fetch('/api/v1/app-icons', {
      method: 'POST',
      body: form,
      credentials: 'same-origin',
      headers: { ...csrfHeaders() },
    })
    if (!response.ok) {
      let body: Record<string, unknown> = {}
      try {
        body = await response.json()
      } catch {
        /* 본문이 JSON이 아니면 상태만 남긴다 */
      }
      throw new PortalApiError(response.status, body)
    }
    return response.json()
  },


  auditLogs: (params: {
    actor_username?: string
    action?: string
    page?: number
    size?: number
  }) => {
    const q = new URLSearchParams()
    for (const [k, v] of Object.entries(params)) if (v) q.set(k, String(v))
    return api.get<Page<AuditLog>>(`/audit-logs?${q}`)
  },
  auditActions: () => api.get<string[]>('/audit-logs/actions'),
}
